presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/ColumnStatsBuilder.java

```diff
@@ -0,0 +1,237 @@
+
+        if (type.equals(BOOLEAN)) {
+            return indexBoolean(type, reader, columnIndex, columnId);
+        }
+        if (type.equals(BIGINT) ||
+                type.equals(INTEGER) ||
+                type.equals(SMALLINT) ||
+                type.equals(TINYINT) ||
+                type.equals(DATE) ||
+                type.equals(TIME) ||
+                type.equals(TIMESTAMP)) {
+            return indexLong(type, reader, columnIndex, columnId, Type::getLong);
+        }
+        if (type.equals(DOUBLE)) {
+            return indexLong(type, reader, columnIndex, columnId, ColumnStatsBuilder::readDouble);
+        }
+        if (type.equals(REAL)) {
+            return indexLong(type, reader, columnIndex, columnId, ColumnStatsBuilder::readReal);
+        }
+        if (isVarcharType(type) || type.equals(VARBINARY)) {
+            return indexBinary(type, reader, columnIndex, columnId);
+        }
+        return null;
+    }
+
+    private static int columnIndex(List<String> columnNames, long columnId)
+    {
+        int index = columnNames.indexOf(String.valueOf(columnId));
+        if (index == -1) {
+            throw new PrestoException(RAPTOR_INTERNAL_ERROR, "Missing column ID: " + columnId);
+        }
+        return index;
+    }
+
+    private static ColumnStats indexBoolean(Type type, OrcRecordReader reader, int columnIndex, long columnId)
+            throws IOException
+    {
+        boolean set = false;
+        boolean min = false;
+        boolean max = false;
+
+        while (true) {
+            int batchSize = reader.nextBatch();
+            if (batchSize <= 0) {
+                break;
+            }
+            Block block = reader.readBlock(type, columnIndex);
+
+            for (int i = 0; i < batchSize; i++) {
+                if (block.isNull(i)) {
+                    continue;
+                }
+                boolean value = type.getBoolean(block, i);
+                if (!set) {
+                    set = true;
+                    min = value;
+                    max = value;
+                }
+                else if (!value) {
+                    min = false;
+                }
+                else {
+                    max = true;
+                }
+            }
+        }
+
+        return new ColumnStats(columnId, set ? min : null, set ? max : null);
+    }
+
+    private static ColumnStats indexLong(Type type, OrcRecordReader reader, int columnIndex, long columnId, ToLong toLong)
+            throws IOException
+    {
+        boolean set = false;
+        long min = 0;
+        long max = 0;
+
+        while (true) {
+            int batchSize = reader.nextBatch();
+            if (batchSize <= 0) {
+                break;
+            }
+            Block block = reader.readBlock(type, columnIndex);
+
+            for (int i = 0; i < batchSize; i++) {
+                if (block.isNull(i)) {
```

Comment:
There is one case worth considering:

table1 bucket on 'userid', temporal='ds';
There are 1000 rows that both 'userid' and 'ds' are all NULL; right now they may be in one chunk, and ColumnStats has 'min' 'max' as null;
So ChunkMetadata.java line 146, throws "null pointer exception". 

If we treat "both 'userid' and 'ds' all NULL " as invalid data, then this is not a logical error. But maybe output accurate error message in here, like "writing into tableXX chunkYY failed, temporal columns are all NULL".

Reply:
Great find! This is indeed a bug. We should fix it in `RaptorPageSink`, since that is where it splits chunks by day. Currently, it is not checking for null, so the date value being read might have arbitrary garbage.